### PR TITLE
JAVA-2260: Handle empty collections in PreparedStatement.bind(...)

### PR DESCRIPTION
**Pull request by olim7t**
Motivation:

`preparedStatement.bind(new HashSet<>())` fails with a
CodecNotFoundException.

This is because the codec registry defaults to Set<Boolean> when it
tries to infer the type of the value (this works fine for lookups by
Java type only, because all empty collections are encoded in the same
way, so any collection codec will do -- but here we happen to also know
the CQL type, and it doesn't match).

Modifications:

Special-case empty collections at the beginning of codecFor(CqlType,
value): simply look up a collection codec for an arbitrary element type.

Result:

We return a codec that can encode the provided value.
Note that it doesn't accept the CQL type, nor the declared Java type of
the value, but the contract of the method doesn't say that it should.

### CODE REVIEW

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java
Comment (tomekl007) [inline core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java:509]: If the type is not important why we don't pass Object?
Or maybe even better we could create marker `class Any` and use it here for a readability

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java
Comment (adutra) [inline core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java]: Two remarks:

1) I don't think this is going to solve the problem for all cases. For example, if you change your unit test to this:

```
    ListType cqlType = DataTypes.listOf(DataTypes.listOf(DataTypes.INT));
    List<List<Integer>> value = Collections.singletonList(Collections.emptyList());
```
The test will fail because the enclosing collection is not empty.
IMO the correct way to handle this is inside `inspectType` and `getCachedCodec`. You might need to give up on generic codecs for empty collections and actually create a codec that truly accepts the target CQL type.

2) I also think that this added code snippet is appearing too soon, and in particular it is bypassing all user-registered codecs, but users might have registered special codecs for collections, in which case we should use them if possible.

File: core/src/test/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistryTest.java
Comment (adutra) [inline core/src/test/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistryTest.java]: Like I said above the tests would be more meaningful if the enclosing collection wasn't empty but the enclosed one was.

**Review by adutra — CHANGES_REQUESTED** [review]
I went ahead and tried to fix the issues I pointed out here in a branch named `java2260-adu`. Waiting for CI now. Let me know what you think.

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java
Comment (olim7t) [reply core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java:509]: Because the registry will still look for a codec for the element type, and there is no `Object` codec. We pick one of the primitive CQL types.

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java
Comment (olim7t) [reply core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java]: Good catch.
I like with your solution of inferring the Java type from the CQL type if available, I'm cherry-picking your commits on this PR.

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java
Comment (olim7t) [inline core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java:317]: IDEA reports a few duplicate branches here, we can factor the cases that return the same result.

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java
Comment (tomekl007) [inline core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java:306]: don't we want to provide some upper bound for that recursion?
Is it possible that the user will create so many enclosing collections that our code will throw `StackOverflow`?

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java
Comment (adutra) [reply core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java:317]: Yeah I noticed them too but couldn't decide what is best: to factor together or to leave the case instructions in the order in which they are declared in `ProtocolConstants.DataType`. Your call.

File: core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java
Comment (adutra) [reply core/src/main/java/com/datastax/oss/driver/internal/core/type/codec/registry/CachingCodecRegistry.java:306]: I think it would blow up server-side first, because you would need to create a collection type with thousands of nested types for that to happen, as in `list<list<list<list<...`.

**Review by adutra — APPROVED** [review]

**Pull request merged**
